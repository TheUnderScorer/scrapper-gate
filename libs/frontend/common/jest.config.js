module.exports = {
  displayName: 'frontend-common',
  preset: '../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../coverage/libs/frontend/common',
};
