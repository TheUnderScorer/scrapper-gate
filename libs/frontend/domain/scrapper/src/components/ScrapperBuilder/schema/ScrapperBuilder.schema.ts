import { ScrapperBuilderFormState } from '../ScrapperBuilder.types';
import {
  ScrapperRunSettingsInputSchema,
  VariableInputSchema,
} from '@scrapper-gate/shared/validation';
import joi from 'joi';
import { ScrapperStepNodeSchema } from './ScrapperStepNode.schema';

export const ScrapperBuilderSchema = joi.object<ScrapperBuilderFormState>({
  items: joi.array().items(ScrapperStepNodeSchema),
  name: joi.string().required(),
  variables: joi.array().items(VariableInputSchema).allow(null),
  runSettings: ScrapperRunSettingsInputSchema,
});
