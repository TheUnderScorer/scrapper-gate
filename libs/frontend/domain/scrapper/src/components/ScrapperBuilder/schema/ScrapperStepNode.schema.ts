import { FlowBuilderNodeTypes } from '@scrapper-gate/frontend/flow-builder';
import { ScrapperStepInputSchema } from '@scrapper-gate/shared/validation';
import joi from 'joi';
import { isEdge } from 'react-flow-renderer';
import { ScrapperBuilderNode } from '../ScrapperBuilder.types';

const omitTypes = [FlowBuilderNodeTypes.Start, FlowBuilderNodeTypes.End];

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const whenSchema = joi.object().custom((item: ScrapperBuilderNode, helpers) => {
  if (isEdge(item) || omitTypes.includes(item.type as FlowBuilderNodeTypes)) {
    return helpers.error('any');
  }

  return item;
});
export const ScrapperStepNodeSchema = joi
  .object<Pick<ScrapperBuilderNode, 'data' | 'type'>>({
    data: joi.object(),
    type: joi.string(),
  })
  .when(whenSchema, {
    then: {
      data: ScrapperStepInputSchema,
    },
  });
