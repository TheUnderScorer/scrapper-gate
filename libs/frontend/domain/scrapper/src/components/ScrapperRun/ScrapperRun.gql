query GetMyScrapperRun($id: ID!) {
  getMyScrapperRun(id: $id) {
    id
    createdAt
    endedAt
    name
    scrapper {
      id
      name
      steps {
        id
      }
    }
    error {
      date
      message
      name
      stepId
    }
    steps {
      ...ScrapperBuilderStep
    }
    state
    error {
      date
      message
      name
    }
    keyPairValues
    results {
      id
      endedAt
      startedAt
      state
      performance {
        duration
      }
      step {
        ...ScrapperBuilderStep
      }
      values {
        id
        sourceElement {
          id
          classNames
          tag
        }
        value
      }
    }
  }
}

query GetMyScrapperRunState($id: ID!) {
  getMyScrapperRun(id: $id) {
    id
    state
    scrapper {
      id
      name
      steps {
        id
      }
      lastRun {
        id
        state
        endedAt
      }
    }
    keyPairValues
    results {
      id
      state
      performance {
        duration
      }
      error {
        date
        message
        name
      }
      step {
        id
        key
      }
      endedAt
      startedAt
      values {
        id
        sourceElement {
          id
          classNames
          tag
        }
        value
      }
    }
  }
}
