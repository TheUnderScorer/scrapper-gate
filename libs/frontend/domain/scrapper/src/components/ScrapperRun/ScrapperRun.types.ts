import { ApolloError } from '@apollo/client';
import { FetchPolicyProps } from '@scrapper-gate/frontend/common';
import {
  BaseNodeProperties,
  FlowBuilderFormState,
  FlowBuilderProps,
} from '@scrapper-gate/frontend/flow-builder';
import { Exists } from '@scrapper-gate/shared/common';
import {
  RouteCreator,
  ScrapperRouteParams,
} from '@scrapper-gate/shared/routing';
import {
  GetMyScrapperRunQuery,
  Scrapper,
  ScrapperStep,
} from '@scrapper-gate/shared/schema';
import { Node } from 'react-flow-renderer';

export interface ScrapperRunProps
  extends Pick<FlowBuilderProps, 'onClose'>,
    FetchPolicyProps {
  runId: string;
  onQueryError?: (error: ApolloError) => unknown;
  scrapperUrlCreator: RouteCreator<ScrapperRouteParams>;
}

export type ScrapperRunStep = Pick<ScrapperStep, 'id'>;

export type ScrapperRunNode = Node<ScrapperRunNodeProperties>;

export type ScrapperRunNodeProperties = BaseNodeProperties & {
  runResult?: Exists<Exists<ScrapperRunFlowBuilder>['results']>[number];
};

export interface ScrapperRunScrapper extends Pick<Scrapper, 'id'> {
  steps?: ScrapperRunStep[];
}

export interface ScrapperRunFormState
  extends FlowBuilderFormState<ScrapperRunNodeProperties> {
  scrapper?: ScrapperRunScrapper;
}

export type ScrapperRunFlowBuilder = GetMyScrapperRunQuery['getMyScrapperRun'];
