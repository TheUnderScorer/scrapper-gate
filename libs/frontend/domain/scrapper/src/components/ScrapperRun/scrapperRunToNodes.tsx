import { Selection } from '@scrapper-gate/frontend/common';
import { ScrapperBuilderNodeSelection } from '../../shared/scrapperNodeSelection';
import { scrapperStepsToNodes } from '../../shared/scrapperStepsToNodes';
import { ScrapperBuilderStep } from '../../shared/types';
import { ScrapperRunNodeAddonBefore } from './NodeAddonBefore/ScrapperRunNodeAddonBefore';
import { ScrapperRunFlowBuilder, ScrapperRunNode } from './ScrapperRun.types';

export const scrapperRunToNodes = (
  run: ScrapperRunFlowBuilder,
  selections: Selection<ScrapperBuilderNodeSelection>[]
) => {
  let steps = (run?.steps ?? []) as Array<
    ScrapperBuilderStep & Partial<ScrapperRunNode>
  >;

  if (run?.results) {
    steps = run.results.map((result) => ({
      ...result.step,
      runResult: result,
      nodeAddonBefore: (node: ScrapperRunNode) => (
        <ScrapperRunNodeAddonBefore node={node} />
      ),
    }));
  }

  return scrapperStepsToNodes({ steps: steps, selections: selections });
};
