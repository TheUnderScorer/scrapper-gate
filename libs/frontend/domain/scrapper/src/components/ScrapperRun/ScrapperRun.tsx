import { ApolloError } from '@apollo/client';
import { Divider, Stack, Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import { submitNotAllowed } from '@scrapper-gate/frontend/common';
import {
  FlowBuilder,
  FlowBuilderTabsSelection,
  flowBuilderUtils,
  NodeContentProps,
} from '@scrapper-gate/frontend/flow-builder';
import { useGetMyScrapperRunQuery } from '@scrapper-gate/frontend/schema';
import { useSnackbarOnError } from '@scrapper-gate/frontend/snackbars';
import {
  ButtonRouteLink,
  ReturnBtn,
  RunState,
  RunStateEntity,
} from '@scrapper-gate/frontend/ui';
import { isCompleted } from '@scrapper-gate/shared/run-states';
import { Maybe } from '@scrapper-gate/shared/schema';
import { FormApi } from 'final-form';
import React, { useCallback, useEffect, useMemo, useRef } from 'react';
import { Form } from 'react-final-form';
import { v4 } from 'uuid';
import { scrapperRunPollMs } from '../../shared/constants';
import { createScrapperNodeSelection } from '../../shared/scrapperNodeSelection';
import { ScrapperRunNodeContent } from './NodeContent/ScrapperRunNodeContent';
import { ScrapperRunFormState, ScrapperRunProps } from './ScrapperRun.types';
import { scrapperRunToNodes } from './scrapperRunToNodes';
import { ScrapperRunValuesTable } from './ValuesTable/ScrapperRunValuesTable';

const PREFIX = 'ScrapperRun';

const classes = {
  divider: `${PREFIX}-divider`,
};

const StyledStack = styled(Stack)(() => ({
  [`& .${classes.divider}`]: {
    height: '30px',
  },
}));

const selection = createScrapperNodeSelection();

const initialNodes = [
  flowBuilderUtils.createStartNode(
    {
      x: 0,
      y: 0,
    },
    true
  ),
];

const handleConnect = flowBuilderUtils.basicHandleConnect();
const handleAdd = flowBuilderUtils.basicHandleAddNode(() => v4());

export const ScrapperRun = ({
  runId,
  onQueryError,
  scrapperUrlCreator,
  fetchPolicy,
  ...rest
}: ScrapperRunProps) => {
  const formRef = useRef<Maybe<FormApi<ScrapperRunFormState>>>();

  const snackbarOnError = useSnackbarOnError();

  const handleError = useCallback(
    (error: ApolloError) => {
      snackbarOnError(error);

      onQueryError?.(error);
    },
    [onQueryError, snackbarOnError]
  );

  const { data, loading, startPolling, stopPolling } = useGetMyScrapperRunQuery(
    {
      variables: {
        id: runId,
      },
      onError: handleError,
      fetchPolicy,
      onCompleted: (data) => {
        if (!isCompleted(data?.getMyScrapperRun?.state)) {
          startPolling(scrapperRunPollMs);
        }
      },
    }
  );

  useEffect(() => {
    if (isCompleted(data?.getMyScrapperRun?.state)) {
      stopPolling();
    }
  }, [data, stopPolling]);

  const nodesCreator = useMemo(
    () =>
      data?.getMyScrapperRun
        ? scrapperRunToNodes(data.getMyScrapperRun, selection)
        : undefined,
    [data]
  );

  // TODO Figure out if data here results in scrapper existing, or if we still need to keep form ref
  const initialValues = useMemo<ScrapperRunFormState>(
    () => ({
      items: initialNodes,
      scrapper: data?.getMyScrapperRun?.scrapper,
    }),
    [data]
  );

  const tabs = useMemo<FlowBuilderTabsSelection[]>(
    () => [
      {
        label: 'Values',
        value: 'values',
        content: (
          <ScrapperRunValuesTable
            values={data?.getMyScrapperRun?.keyPairValues}
          />
        ),
      },
    ],
    [data]
  );

  const nodeContent = useCallback(
    (props: NodeContentProps) => (
      <ScrapperRunNodeContent
        {...props}
        scrapperUrlCreator={scrapperUrlCreator}
      />
    ),
    [scrapperUrlCreator]
  );

  useEffect(() => {
    if (data?.getMyScrapperRun?.scrapper) {
      formRef.current?.change('scrapper', data.getMyScrapperRun.scrapper);
    }
  }, [data]);

  return (
    <Form<ScrapperRunFormState>
      initialValues={initialValues}
      onSubmit={submitNotAllowed}
      render={({ form }) => {
        formRef.current = form;

        return (
          <FlowBuilder
            tabs={tabs}
            onAdd={handleAdd}
            onConnect={handleConnect}
            loading={loading}
            nodesCreator={nodesCreator}
            nodesSelection={selection}
            nodeContentTitle="Step result"
            nodeKeyProperty="data.key"
            defaultNodeContent={nodeContent}
            readOnly
            title={
              <StyledStack direction="row" spacing={2} alignItems="center">
                <ReturnBtn />
                <span>
                  <Typography variant="h6">
                    {data?.getMyScrapperRun?.name}
                  </Typography>
                  {data?.getMyScrapperRun?.scrapper && (
                    <Typography>
                      from{' '}
                      <ButtonRouteLink
                        to={scrapperUrlCreator({
                          scrapperId: data.getMyScrapperRun.scrapper.id,
                        })}
                      >
                        {data.getMyScrapperRun.scrapper.name}
                      </ButtonRouteLink>
                    </Typography>
                  )}
                </span>
                <Divider orientation="vertical" className={classes.divider} />
                {data?.getMyScrapperRun && (
                  <Typography>
                    <RunState
                      showIcon
                      runMutationCalled
                      state={data.getMyScrapperRun.state}
                      entity={RunStateEntity.Scrapper}
                      entityName={data.getMyScrapperRun.name ?? ''}
                    />
                  </Typography>
                )}
              </StyledStack>
            }
            {...rest}
          />
        );
      }}
    />
  );
};
