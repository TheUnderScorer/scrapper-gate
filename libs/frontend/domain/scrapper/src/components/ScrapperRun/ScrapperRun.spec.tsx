/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { QueryParamProvider } from '@scrapper-gate/frontend/common';
import { GetMyScrapperRunDocument } from '@scrapper-gate/frontend/schema';
import { SnackbarProvider } from '@scrapper-gate/frontend/snackbars';
import { ThemeProvider } from '@scrapper-gate/frontend/theme';
import {
  createMockScrapper,
  createMockScrapperRun,
  createMockScrapperStep,
} from '@scrapper-gate/shared/domain/scrapper/mocks';
import {
  GetMyScrapperRunQueryVariables,
  RunState,
  ScrapperAction,
  ScrapperRun as Run,
} from '@scrapper-gate/shared/schema';
import '@testing-library/jest-dom';
import { act, render, RenderResult, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { XYPosition } from 'react-flow-renderer';
import { BrowserRouter } from 'react-router-dom';
import { ScrapperRun } from './ScrapperRun';

async function createRun() {
  const scrapper = createMockScrapper();

  const basePosition: XYPosition = {
    x: 50,
    y: 0,
  };

  const steps = await Promise.all([
    createMockScrapperStep({
      intercept: (step) => {
        step.action = ScrapperAction.Click;
        step.isFirst = true;

        return step;
      },
    }),

    createMockScrapperStep({
      intercept: (step) => {
        step.action = ScrapperAction.ReadText;

        return step;
      },
    }),

    createMockScrapperStep({
      intercept: (step) => {
        step.action = ScrapperAction.Condition;

        return step;
      },
    }),

    createMockScrapperStep({
      intercept: (step) => {
        step.action = ScrapperAction.Type;

        return step;
      },
    }),

    createMockScrapperStep({
      intercept: (step) => {
        step.action = ScrapperAction.ReloadPage;

        return step;
      },
    }),
  ]);

  steps[0].nextStep = steps[1];
  steps[1].nextStep = steps[2];
  steps[2].stepOnTrue = steps[3];
  steps[2].stepOnFalse = steps[4];

  steps.forEach((step, index) => {
    step.position = {
      ...basePosition,
      x: basePosition.x * index,
    };
  });

  const run = createMockScrapperRun(steps);

  run.scrapper = scrapper;
  run.state = RunState.Pending;

  return run;
}

let mocks: MockedResponse[];
let run: Run;

async function renderCmp() {
  let cmp: RenderResult;

  act(() => {
    cmp = render(
      <BrowserRouter>
        <QueryParamProvider>
          <SnackbarProvider>
            <ThemeProvider>
              <MockedProvider mocks={mocks}>
                <ScrapperRun
                  runId={run.id}
                  scrapperUrlCreator={jest.fn()}
                  fetchPolicy="no-cache"
                />
              </MockedProvider>
            </ThemeProvider>
          </SnackbarProvider>
        </QueryParamProvider>
      </BrowserRouter>
    );
  });

  return cmp!;
}

async function waitForNodes(cmp: RenderResult) {
  await waitFor(() => {
    const nodes = cmp.container.querySelectorAll('.flow-builder-node');

    // All steps + start node
    expect(nodes).toHaveLength(run.steps!.length + 1);
  });
}

describe('<ScrapperRun />', () => {
  beforeEach(async () => {
    run = await createRun();

    mocks = [
      {
        request: {
          query: GetMyScrapperRunDocument,
          variables: {
            id: run.id,
          } as GetMyScrapperRunQueryVariables,
        },
        result: {
          data: {
            getMyScrapperRun: run,
          },
        },
      },
    ];
  });

  it('should render without crashing', async () => {
    const cmp = await renderCmp();

    expect(cmp).toMatchSnapshot();
  });

  it('should render steps correctly', async () => {
    const cmp = await renderCmp();

    await waitForNodes(cmp);
  });

  it('should indicate if step is running', async () => {
    run.results![0].state = RunState.InProgress;

    const cmp = await renderCmp();

    await waitForNodes(cmp);

    const element = cmp.container.querySelector(
      `#node-${run.results![0].step.id} .scrapper-run-state`
    );

    expect(element).toBeInTheDocument();
  });

  it('should render run data', async () => {
    const result = run.results![1];

    const cmp = await renderCmp();

    await waitForNodes(cmp);

    const target = cmp.container.querySelector(
      `#node-${result.step.id} .action-node-box`
    );

    act(() => {
      userEvent.dblClick(target!);
    });

    const content = await cmp.container.querySelector(
      '.scrapper-run-node-content'
    );

    expect(content).toBeInTheDocument();
  });
});
