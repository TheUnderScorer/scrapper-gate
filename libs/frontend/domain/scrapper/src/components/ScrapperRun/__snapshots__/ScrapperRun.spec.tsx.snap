// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScrapperRun /> should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded makeStyles-paper-3 css-172kgz0-MuiPaper-root"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionStatic css-pivuyh-MuiPaper-root-MuiAppBar-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular makeStyles-appBar-5 css-1t29gy6-MuiToolbar-root"
          >
            <div
              class="makeStyles-mainGrid-6 css-hz4k7x-Stack-root"
            >
              <div
                class="css-1jz9r7a-Stack-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-fomgco-MuiSkeleton-root"
                  style="width: 30px; height: 30px;"
                />
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                  style="width: 60px; height: 10px;"
                />
              </div>
              <div
                class="css-1jz9r7a-Stack-root"
              />
            </div>
          </div>
        </header>
        <div
          class="MuiPaper-root MuiPaper-outlined makeStyles-paper-10 css-mta8o9-MuiPaper-root"
        >
          <div
            class="MuiTabs-root makeStyles-tabs-9 css-1ujnqem-MuiTabs-root"
          >
            <div
              class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
              style="overflow: hidden; margin-bottom: 0px;"
            >
              <div
                class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                role="tablist"
              >
                <button
                  aria-selected="true"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-e0qzyo-MuiButtonBase-root-MuiTab-root"
                  role="tab"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiTab-wrapper css-17yibpu-MuiTab-wrapper"
                  >
                    Steps
                  </span>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-selected="false"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-e0qzyo-MuiButtonBase-root-MuiTab-root"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiTab-wrapper css-17yibpu-MuiTab-wrapper"
                  >
                    Values
                  </span>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <span
                class="MuiTabs-indicator css-78exr-MuiTabs-indicator"
                style="left: 0px; width: 0px;"
              />
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1x5kfkn"
        >
          <div
            class="makeStyles-stack-11 css-1abmola-Stack-root"
          >
            <div
              class="centered makeStyles-paper-12 MuiBox-root css-1rr92ez"
            >
              <div
                class="css-m72s89-Stack-root"
              >
                <span
                  class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1h9nrm7-MuiCircularProgress-root"
                  role="progressbar"
                  style="width: 40px; height: 40px;"
                >
                  <svg
                    class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                      cx="44"
                      cy="44"
                      fill="none"
                      r="20.2"
                      stroke-width="3.6"
                    />
                  </svg>
                </span>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-108qaiy-MuiTypography-root"
                >
                  Loading builder...
                </p>
              </div>
            </div>
            <div
              class="makeStyles-drawer-16 node-content-closed makeStyles-container-22"
              style="position: relative; user-select: auto; visibility: hidden; width: 20%; height: 100%; max-width: 900px; min-width: 500px; box-sizing: border-box; flex-shrink: 0;"
            >
              <div
                class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 makeStyles-paper-18 makeStyles-paper-20 makeStyles-content-23 resizable-panel-content css-1ps6pg7-MuiPaper-root"
                style="webkit-transform: translateX(1024px) translateX(0px); transform: translateX(1024px) translateX(0px);"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap makeStyles-container-26 layout css-19kb62l-MuiGrid-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item makeStyles-body-27 makeStyles-body-28 layout-body css-10texkh-MuiGrid-root"
                    style="height: calc(100% - 50px); flex: 1;"
                  >
                    <div
                      class="makeStyles-content-14 node-content css-nu13me-Stack-root"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-h6 css-5d68dr-MuiTypography-root"
                      >
                        Step result
                      </h6>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item layout-footer css-10texkh-MuiGrid-root"
                    style="height: 50px;"
                  >
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth makeStyles-btnDivider-15 css-9mgopn-MuiDivider-root"
                    />
                    <div
                      class="MuiBox-root css-1v7vvra"
                    >
                      <button
                        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButtonBase-root Mui-focusVisible css-1b9y47s-MuiButtonBase-root-MuiButton-root"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label css-8xplcm-MuiButton-label"
                        >
                          Close
                        </span>
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        >
                          <span
                            class="css-1yvx2br-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                            style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
                          >
                            <span
                              class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                            />
                          </span>
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="makeStyles-handle-21 makeStyles-handle-25"
              >
                <div
                  class=""
                  style="position: absolute; user-select: none; width: 10px; height: 100%; top: 0px; left: -5px; cursor: col-resize;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded makeStyles-paper-3 css-172kgz0-MuiPaper-root"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionStatic css-pivuyh-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular makeStyles-appBar-5 css-1t29gy6-MuiToolbar-root"
        >
          <div
            class="makeStyles-mainGrid-6 css-hz4k7x-Stack-root"
          >
            <div
              class="css-1jz9r7a-Stack-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-fomgco-MuiSkeleton-root"
                style="width: 30px; height: 30px;"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                style="width: 60px; height: 10px;"
              />
            </div>
            <div
              class="css-1jz9r7a-Stack-root"
            />
          </div>
        </div>
      </header>
      <div
        class="MuiPaper-root MuiPaper-outlined makeStyles-paper-10 css-mta8o9-MuiPaper-root"
      >
        <div
          class="MuiTabs-root makeStyles-tabs-9 css-1ujnqem-MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-e0qzyo-MuiButtonBase-root-MuiTab-root"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper css-17yibpu-MuiTab-wrapper"
                >
                  Steps
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-e0qzyo-MuiButtonBase-root-MuiTab-root"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiTab-wrapper css-17yibpu-MuiTab-wrapper"
                >
                  Values
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-78exr-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-1x5kfkn"
      >
        <div
          class="makeStyles-stack-11 css-1abmola-Stack-root"
        >
          <div
            class="centered makeStyles-paper-12 MuiBox-root css-1rr92ez"
          >
            <div
              class="css-m72s89-Stack-root"
            >
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1h9nrm7-MuiCircularProgress-root"
                role="progressbar"
                style="width: 40px; height: 40px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 css-108qaiy-MuiTypography-root"
              >
                Loading builder...
              </p>
            </div>
          </div>
          <div
            class="makeStyles-drawer-16 node-content-closed makeStyles-container-22"
            style="position: relative; user-select: auto; visibility: hidden; width: 20%; height: 100%; max-width: 900px; min-width: 500px; box-sizing: border-box; flex-shrink: 0;"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 makeStyles-paper-18 makeStyles-paper-20 makeStyles-content-23 resizable-panel-content css-1ps6pg7-MuiPaper-root"
              style="webkit-transform: translateX(1024px) translateX(0px); transform: translateX(1024px) translateX(0px);"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap makeStyles-container-26 layout css-19kb62l-MuiGrid-root"
              >
                <div
                  class="MuiGrid-root MuiGrid-item makeStyles-body-27 makeStyles-body-28 layout-body css-10texkh-MuiGrid-root"
                  style="height: calc(100% - 50px); flex: 1;"
                >
                  <div
                    class="makeStyles-content-14 node-content css-nu13me-Stack-root"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-h6 css-5d68dr-MuiTypography-root"
                    >
                      Step result
                    </h6>
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item layout-footer css-10texkh-MuiGrid-root"
                  style="height: 50px;"
                >
                  <hr
                    class="MuiDivider-root MuiDivider-fullWidth makeStyles-btnDivider-15 css-9mgopn-MuiDivider-root"
                  />
                  <div
                    class="MuiBox-root css-1v7vvra"
                  >
                    <button
                      class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButtonBase-root Mui-focusVisible css-1b9y47s-MuiButtonBase-root-MuiButton-root"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="MuiButton-label css-8xplcm-MuiButton-label"
                      >
                        Close
                      </span>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      >
                        <span
                          class="css-1yvx2br-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                          style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
                        >
                          <span
                            class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                          />
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="makeStyles-handle-21 makeStyles-handle-25"
            >
              <div
                class=""
                style="position: absolute; user-select: none; width: 10px; height: 100%; top: 0px; left: -5px; cursor: col-resize;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
